### PR TITLE
Recalculate first places on user restriction/unrestriction

### PR DESCRIPTION
**Pull request by RealistikDash**
This PR adds functionality to the `peppy:ban` Redis pubsub, where it will proceed to recalculate the first places for all the beatmaps the user has first places on upon restriction, or try to restore their first places upon unrestriction.

Requires testing.

### CODE REVIEW

File: app/usecases/score.py [hunk 1]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+        return
+
+    leaderboard = await app.usecases.leaderboards.fetch(beatmap, mode)
+    unrestricted_scores = await leaderboard.get_unrestricted_scores(0)
+
+    if cur_first is not None:
+        # No more scores on the map.
+        if not unrestricted_scores:
+            await delete_first_place(cur_first["scoreid"])
+            return
```
Comment (tsunyoku) [inline app/usecases/score.py]: remove

File: app/usecases/score.py [hunk 2]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+        else:
+            lb_first = unrestricted_scores[0]
+            # First place is unchanged.
+            if lb_first.id == cur_first["scoreid"]:
+                return
```
Comment (tsunyoku) [inline app/usecases/score.py:273]: new line after this pls

File: app/usecases/score.py [hunk 3]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+async def get_user_score_maps(user_id: int, mode: Mode) -> list[str]:
+    beatmaps_db = await app.state.services.read_database.fetch_all(
+        f"SELECT beatmap_md5 FROM {mode.scores_table} WHERE completed = 3 "
+        "AND play_mode = :mode AND userid = :user_id",
+        {
+            "mode": mode.as_vn,
+            "user_id": user_id,
+        },
+    )
+    return [row[0] for row in beatmaps_db]
```
Comment (tsunyoku) [inline app/usecases/score.py]: tuple is sin

File: app/usecases/score.py [hunk 4]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+        },
+    )
+
+    # Calculate the top on the leaderboards. Use leaderboard as its likely to be cached.
+    beatmap = await app.usecases.beatmap.fetch_by_md5(beatmap_md5)
+    if beatmap is None:
+        return
+
+    leaderboard = await app.usecases.leaderboards.fetch(beatmap, mode)
+    unrestricted_scores = await leaderboard.get_unrestricted_scores(0)
```
Comment (tsunyoku) [inline app/usecases/score.py]: use kwarg

File: app/usecases/score.py [hunk 5]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+            "mode": score.mode.as_vn,
+            "rx": score.mode.relax_int,
+            "sid": score.id,
+            "uid": score.user_id,
+        },
+    )
+
+
+async def calculate_first_place(beatmap_md5: str, mode: Mode) -> None:
+    cur_first = await app.state.services.read_database.fetch_one(
```
Comment (tsunyoku) [inline app/usecases/score.py]: `cur`?

File: app/usecases/score.py [hunk 6]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
         .write_raw(raw_data)
         .write_i64_le(score.id)
     )
+
+
+class FirstPlaceRow(NamedTuple):
```
Comment (tsunyoku) [inline app/usecases/score.py]: would kinda like TypedDict to be consistent with rest of codebase but meh

File: app/usecases/score.py [hunk 7]
```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
...
+            lb_first = unrestricted_scores[0]
+            # First place is unchanged.
+            if lb_first.id == cur_first["scoreid"]:
+                return
+            await delete_first_place(cur_first["scoreid"])
+            await insert_first_place(lb_first)
+
+
+async def get_user_score_maps(user_id: int, mode: Mode) -> list[str]:
+    beatmaps_db = await app.state.services.read_database.fetch_all(
```
Comment (tsunyoku) [inline app/usecases/score.py]: new line after this pls

**Review by tsunyoku — CHANGES_REQUESTED** [review]

File: app/usecases/score.py (hunk 1 above)
Comment (RealistikDash) [reply app/usecases/score.py]: Fixed

File: app/usecases/score.py (hunk 2 above)
Comment (RealistikDash) [reply app/usecases/score.py:273]: Fixed

File: app/usecases/score.py (hunk 3 above)
Comment (RealistikDash) [reply app/usecases/score.py]: Fixed

File: app/usecases/score.py (hunk 4 above)
Comment (RealistikDash) [reply app/usecases/score.py]: Fixed

File: app/usecases/score.py (hunk 5 above)
Comment (RealistikDash) [reply app/usecases/score.py]: Renamed to `current_first`, should be clearer

File: app/usecases/score.py (hunk 6 above)
Comment (RealistikDash) [reply app/usecases/score.py]: changed to `TypedDict`

File: app/usecases/score.py (hunk 7 above)
Comment (RealistikDash) [reply app/usecases/score.py]: Fixed

**Comment by cmyui** [comment]
will check this out soon - looks good

File: app/objects/leaderboard.py [hunk 8]
```diff
@@ -37,38 +37,39 @@ async def find_user_score(
         unrestricted: bool = True,
     ) -> Optional[UserScore]:
         if unrestricted:
-            unrestricted_scores = await self.get_unrestricted_scores(user_id)
+            scores = await self.get_scores(restricted_users=[user_id])
```
Comment (cmyui) [inline app/objects/leaderboard.py]: this is how i ended up refactoring this function we chatted about earlier

**Review by cmyui — CHANGES_REQUESTED** [review]
this works for restrictions from outside of the score service via the pubsub, but i don't seem to see this hooked up to (for example) the restrict_user usecase that's used throughout this codebase

that's used in other cases like e.g. when a user is restricted for submitting a score above the pp cap, or using an outdated leaderboard version

**Comment by cmyui** [comment]
(ideally, the high level redis pubsubs should be treated the same as api endpoints and should use the usecases rather than containing logic in themselves)

**Comment by tsunyoku** [comment]
> this works for restrictions from outside of the score service via the pubsub, but i don't seem to see this hooked up to (for example) the restrict_user usecase that's used throughout this codebase
> 
> that's used in other cases like e.g. when a user is restricted for submitting a score above the pp cap, or using an outdated leaderboard version

doesn’t restrict_user publish to `peppy:ban`?

**Comment by cmyui** [comment]
> doesn’t restrict_user publish to `peppy:ban`?

oh whaatttttt

**Comment by tsunyoku** [comment]
> > doesn’t restrict_user publish to `peppy:ban`?
> 
> oh whaatttttt

https://github.com/osuAkatsuki/score-service/blob/master/app/usecases/user.py#L216

**Comment by cmyui** [comment]
it is real pr salvaging hours

**Comment by tsunyoku** [comment]
bancho-service is already doing this

**Pull request closed without merging**